File: algo.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <random>
#include <cmath>
#include <set>
#include <limits>
#include <algorithm>

class AntColony
{
private:
  int n_ants;
  int n_iterations;
  double decay;
  double alpha;
  double beta;
  std::mt19937 rng;

public:
  AntColony(int ants, int iterations, double decay_rate, double a = 1, double b = 2)
      : n_ants(ants), n_iterations(iterations), decay(decay_rate), alpha(a), beta(b)
  {
    std::random_device rd;
    rng = std::mt19937(rd());
  }

  std::vector<std::vector<double>> build_graph(const std::string &filename)
  {
    std::ifstream file(filename);
    int v1, v2, weight;
    int max_vertex = 0;

    // Первый проход для определения размера матрицы
    while (file >> v1 >> v2 >> weight)
    {
      max_vertex = std::max(max_vertex, std::max(v1, v2));
    }

    // Создаем матрицу расстояний
    int n = max_vertex + 1;
    std::vector<std::vector<double>> distances(n, std::vector<double>(n, std::numeric_limits<double>::infinity()));

    // Второй проход для заполнения матрицы
    file.clear();
    file.seekg(0);
    while (file >> v1 >> v2 >> weight)
    {
      distances[v1][v2] = weight;
      distances[v2][v1] = weight;
    }

    // Заполняем отсутствующие рёбра большими весами
    double max_weight = 0;
    for (const auto &row : distances)
    {
      for (double weight : row)
      {
        if (weight != std::numeric_limits<double>::infinity())
        {
          max_weight = std::max(max_weight, weight);
        }
      }
    }
    max_weight *= 10;

    for (auto &row : distances)
    {
      for (double &weight : row)
      {
        if (weight == std::numeric_limits<double>::infinity())
        {
          weight = max_weight;
        }
      }
    }

    return distances;
  }

  double calculate_path_length(const std::vector<int> &path, const std::vector<std::vector<double>> &distances)
  {
    double length = 0;
    for (size_t i = 0; i < path.size(); ++i)
    {
      length += distances[path[i]][path[(i + 1) % path.size()]];
    }
    std::cout << length << std::endl;
    return length;
  }

  std::vector<double> calculate_probabilities(int current, const std::set<int> &unvisited,
                                              const std::vector<std::vector<double>> &distances,
                                              const std::vector<std::vector<double>> &pheromone)
  {
    std::vector<double> probabilities;
    for (int city : unvisited)
    {
      double p = std::pow(pheromone[current][city], alpha) * std::pow(1.0 / distances[current][city], beta);
      probabilities.push_back(p);
    }

    double total = 0;
    for (double p : probabilities)
    {
      total += p;
    }

    if (total == 0)
    {
      probabilities = std::vector<double>(unvisited.size(), 1.0 / unvisited.size());
    }
    else
    {
      for (double &p : probabilities)
      {
        p /= total;
      }
    }

    return probabilities;
  }

  std::vector<int> construct_path(const std::vector<std::vector<double>> &distances,
                                  const std::vector<std::vector<double>> &pheromone)
  {
    int n_cities = distances.size();
    std::set<int> unvisited;
    for (int i = 0; i < n_cities; ++i)
    {
      unvisited.insert(i);
    }

    std::uniform_int_distribution<int> dist(0, n_cities - 1);
    int start = dist(rng);
    std::vector<int> path = {start};
    unvisited.erase(start);

    while (!unvisited.empty())
    {
      int current = path.back();
      std::vector<double> probabilities = calculate_probabilities(current, unvisited, distances, pheromone);

      std::discrete_distribution<int> distribution(probabilities.begin(), probabilities.end());
      std::vector<int> unvisited_vec(unvisited.begin(), unvisited.end());
      int next_city_idx = distribution(rng);
      int next_city = unvisited_vec[next_city_idx];

      path.push_back(next_city);
      unvisited.erase(next_city);
    }

    return path;
  }

  std::vector<std::vector<double>> update_pheromone(std::vector<std::vector<double>> pheromone,
                                                    const std::vector<std::vector<int>> &paths,
                                                    const std::vector<std::vector<double>> &distances)
  {
    // Испарение феромона
    for (auto &row : pheromone)
    {
      for (double &p : row)
      {
        p *= (1.0 - decay);
      }
    }

    // Добавление нового феромона
    for (const auto &path : paths)
    {
      double path_length = calculate_path_length(path, distances);
      for (size_t i = 0; i < path.size(); ++i)
      {
        int current = path[i];
        int next = path[(i + 1) % path.size()];
        pheromone[current][next] += 1.0 / path_length;
        pheromone[next][current] += 1.0 / path_length;
      }
    }

    return pheromone;
  }

  std::pair<std::vector<int>, double> solve(const std::vector<std::vector<double>> &distances)
  {
    int n_cities = distances.size();
    std::vector<std::vector<double>> pheromone(n_cities, std::vector<double>(n_cities, 1.0 / n_cities));
    std::vector<int> best_path;
    double best_length = std::numeric_limits<double>::infinity();
    std::ofstream out_file("of.txt");

    for (int iteration = 0; iteration < n_iterations; ++iteration)
    {
      std::vector<std::vector<int>> paths;
      for (int ant = 0; ant < n_ants; ++ant)
      {
        paths.push_back(construct_path(distances, pheromone));
      }

      double current_best_length = std::numeric_limits<double>::infinity();
      std::vector<int> current_best_path;

      for (const auto &path : paths)
      {
        double length = calculate_path_length(path, distances);
        if (length < current_best_length)
        {
          current_best_length = length;
          current_best_path = path;
        }
      }

      if (current_best_length < best_length)
      {
        best_length = current_best_length;
        best_path = current_best_path;
      }

      out_file << current_best_length << std::endl;
      pheromone = update_pheromone(pheromone, paths, distances);
    }

    return {best_path, best_length};
  }
};

int main()
{
  int n_ants = 20;
  int n_iterations = 100;
  double decay = 0.1;
  double alpha = 1;
  double beta = 2;

  AntColony aco(n_ants, n_iterations, decay, alpha, beta);
  auto distances = aco.build_graph("if.txt");
  auto [best_path, best_length] = aco.solve(distances);

  std::cout << "Best path length: " << best_length << std::endl;
  std::cout << "Best path: ";
  for (int city : best_path)
  {
    std::cout << city << " ";
  }
  std::cout << std::endl;

  return 0;
}

File: algo.py
import numpy as np
from collections import defaultdict
import random
import matplotlib.pyplot as plt

class AntColony:
    def __init__(self, n_ants, n_iterations, decay, alpha=1, beta=2):
        self.n_ants = n_ants
        self.n_iterations = n_iterations
        self.decay = decay
        self.alpha = alpha  # влияние феромона
        self.beta = beta    # влияние расстояния
        
    def build_graph(self, filename):
      # Читаем данные из файла и строим граф
      edges = defaultdict(dict)
      max_vertex = 0  # Добавляем отслеживание максимального номера вершины
      
      with open(filename, 'r') as f:
          for line in f:
              v1, v2, weight = map(int, line.strip().split())
              edges[v1][v2] = weight
              edges[v2][v1] = weight
              max_vertex = max(max_vertex, v1, v2)  # Обновляем максимальный номер вершины
      
      # Создаем матрицу расстояний с размером (max_vertex + 1) × (max_vertex + 1)
      n = max_vertex + 1
      distances = np.full((n, n), float('inf'))
      
      # Заполняем известные расстояния
      for v1 in edges:
          for v2, weight in edges[v1].items():
              distances[v1][v2] = weight
      
      # Достраиваем граф до полного, используя большие веса для отсутствующих ребер
      max_weight = np.max(distances[distances != float('inf')]) * 10
      distances[distances == float('inf')] = max_weight
      
      return distances
        
    def solve(self, distances):
        n_cities = len(distances)
        pheromone = np.ones((n_cities, n_cities)) / n_cities
        best_path = None
        best_path_length = float('inf')
        all_lengths = []
        
        for iteration in range(self.n_iterations):
            paths = self.construct_solutions(distances, pheromone)
            current_best_path = None
            current_best_length = float('inf')
            
            # Находим лучший путь в текущей итерации
            for path in paths:
                path_length = self.calculate_path_length(path, distances)
                if path_length < current_best_length:
                    current_best_length = path_length
                    current_best_path = path
                    
            # Обновляем лучшее глобальное решение
            if current_best_length < best_path_length:
                best_path_length = current_best_length
                best_path = current_best_path
                
            # Записываем длину текущего лучшего пути
            with open('of.txt', 'a') as f:
                f.write(f"{current_best_length}\n")
            all_lengths.append(current_best_length)
                
            # Обновление феромонов
            pheromone = self.update_pheromone(pheromone, paths, distances)
            
        return best_path, best_path_length, all_lengths
    
    def construct_solutions(self, distances, pheromone):
        n_cities = len(distances)
        paths = []
        
        for ant in range(self.n_ants):
            path = self.construct_path(distances, pheromone, n_cities)
            paths.append(path)
            
        return paths
    
    def construct_path(self, distances, pheromone, n_cities):
        unvisited = set(range(n_cities))
        start = random.randint(0, n_cities-1)
        path = [start]
        unvisited.remove(start)
        
        while unvisited:
            current = path[-1]
            probabilities = self.calculate_probabilities(current, unvisited, distances, pheromone)
            next_city = random.choices(list(unvisited), probabilities)[0]
            path.append(next_city)
            unvisited.remove(next_city)
            
        return path
    
    def calculate_probabilities(self, current, unvisited, distances, pheromone):
        probabilities = []
        
        for city in unvisited:
            # Используем феромоны и эвристическую информацию
            p = pow(pheromone[current][city], self.alpha) * pow(1.0/distances[current][city], self.beta)
            probabilities.append(p)
            
        # Нормализация вероятностей
        total = sum(probabilities)
        if total == 0:
            # Если все вероятности нулевые, используем равномерное распределение
            return [1.0/len(unvisited)] * len(unvisited)
        return [p/total for p in probabilities]
    
    def calculate_path_length(self, path, distances):
        length = 0
        for i in range(len(path)):
            length += distances[path[i]][path[(i+1)%len(path)]]
        print(length)
        return length
    
    def update_pheromone(self, pheromone, paths, distances):
        # Испарение феромона
        pheromone *= (1.0 - self.decay)
        
        # Добавление нового феромона
        for path in paths:
            path_length = self.calculate_path_length(path, distances)
            for i in range(len(path)):
                current = path[i]
                next_city = path[(i+1)%len(path)]
                pheromone[current][next_city] += 1.0 / path_length
                pheromone[next_city][current] += 1.0 / path_length
                
        return pheromone

def plot_convergence(lengths):
    plt.figure(figsize=(10, 6))
    plt.plot(lengths)
    plt.title('Convergence of Ant Colony Optimization')
    plt.xlabel('Iteration')
    plt.ylabel('Path Length')
    plt.grid(True)
    plt.savefig('convergence.png')
    plt.close()

# Параметры алгоритма
n_ants = 20
n_iterations = 100
decay = 0.1
alpha = 1
beta = 2

# Очищаем файл с результатами
open('of.txt', 'w').close()

# Создаем экземпляр класса и решаем задачу
aco = AntColony(n_ants, n_iterations, decay, alpha, beta)
distances = aco.build_graph('if.txt')
best_path, best_length, all_lengths = aco.solve(distances)

print(f"Best path length: {best_length}")
print(f"Best path: {best_path}")

# Строим график сходимости
plot_convergence(all_lengths)